src/lens.jl

```diff
@@ -146,8 +146,25 @@ struct IndexLens{I <: Tuple} <: Lens
 end
 
 get(l::IndexLens, obj) = getindex(obj, l.indices...)
-set(l::IndexLens, obj, val) = setindex(obj, val, l.indices...)
+set(l::IndexLens, obj, val) = _setindex(obj, val, l.indices...)
 
+@generated function _setindex(obj, val, indices...)
+    if hasmethod(Base.setindex, Tuple{obj, val, indices...})
```

Comment:
I know `hasmethod` is not pure and therefore is not legit inside `@generated` in principle.  But I've been seeing code deviating from the principle and somehow magically working.  My _guess_ was that this implementation is OK as long as `_setindex` is called after `Base.setindex(obj, ...)` is defined.  Since our StaticArrays tests are passing, this implementation is fine?

Reply:
Can you show a place, where people use code like this? Generally such things can cause bugs depending on import order of packages. But I don't understand enough to judge how safe this case is. I would guess that you could break this by
```julia
using Setfield
using StaticArrays

# define your own StaticArray with custom setindex
```
OTOH if you don't make this a generated function, with the new compiler it might still perform very well. Have you tested?

Reply:
No I don't know the working code base using exactly this.  I should test this and also do benchmark for the dynamic version.

BTW, what is the new compiler optimization you are talking about?  The Union-splitting https://julialang.org/blog/2018/08/union-splitting sounds very close but the branching information still has to sit at the type level, right?  Also, this blog post is mentioning that method lookup is still slow.

But even if `@generated` approach does not work, we can always have a dynamic version as a fallback and then manually implement fast non-branching version for popular types (like `Array`, `Tuple` and `SArray`).  Probably it is not hard to get > 90% coverage by implementing it for a few types.

Reply:
I had better constant propagation in mind. And I thought `hasmethod` might compile down to `true` or `false`, but it does not. It does dynamically look if there is a method.